cirq/protocols/json.py

```diff
@@ -0,0 +1,191 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from typing import Union, Any, Dict, Optional, List, Callable, Type, cast, \
+    TYPE_CHECKING
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    # mypy needs these explicitly imported
+    import cirq.ops.pauli_gates
+    import cirq.devices.unconstrained_device
+
+
+class _ResolverCache:
+    """Lazily import and build registry to avoid circular imports."""
+
+    def __init__(self):
+        self._crd = None
+
+    @property
+    def cirq_class_resolver_dictionary(self) -> Dict[str, Type]:
+        if self._crd is None:
+            import cirq
+            self._crd = {
+                'CCXPowGate': cirq.CCXPowGate,
+                'CCZPowGate': cirq.CCZPowGate,
+                'CNotPowGate': cirq.CNotPowGate,
+                'CSwapGate': cirq.CSwapGate,
+                'CZPowGate': cirq.CZPowGate,
+                'Circuit': cirq.Circuit,
+                'FSimGate': cirq.FSimGate,
+                'GateOperation': cirq.GateOperation,
+                'GlobalPhaseOperation': cirq.GlobalPhaseOperation,
+                'GridQubit': cirq.GridQubit,
+                'HPowGate': cirq.HPowGate,
+                'ISwapPowGate': cirq.ISwapPowGate,
+                'IdentityGate': cirq.IdentityGate,
+                'LineQubit': cirq.LineQubit,
+                'MeasurementGate': cirq.MeasurementGate,
+                'Moment': cirq.Moment,
+                'NamedQubit': cirq.NamedQubit,
+                '_PauliX': cirq.ops.pauli_gates._PauliX,
+                '_PauliY': cirq.ops.pauli_gates._PauliY,
+                '_PauliZ': cirq.ops.pauli_gates._PauliZ,
+                'PauliString': cirq.PauliString,
+                'PhasedXPowGate': cirq.PhasedXPowGate,
+                'SingleQubitPauliStringGateOperation':
+                cirq.SingleQubitPauliStringGateOperation,
+                'SwapPowGate': cirq.SwapPowGate,
+                '_UnconstrainedDeviceType':
+                cirq.devices.unconstrained_device._UnconstrainedDeviceType,
+                'XPowGate': cirq.XPowGate,
+                'XXPowGate': cirq.XXPowGate,
+                'YPowGate': cirq.YPowGate,
+                'YYPowGate': cirq.YYPowGate,
+                'ZPowGate': cirq.ZPowGate,
+                'ZZPowGate': cirq.ZZPowGate,
+
+                # not a cirq class, but treated as one:
+                'complex': complex,
+            }
+        return self._crd
+
+
+RESOLVER_CACHE = _ResolverCache()
+
+
+def _cirq_class_resolver(cirq_type: str) -> Union[None, Type]:
+    return RESOLVER_CACHE.cirq_class_resolver_dictionary.get(cirq_type, None)
+
+
+DEFAULT_RESOLVERS = [
```

Comment:
What's the purpose of having this as a public constant?  So users can add to it their own resolvers?  If so this should be documented.

Reply:
That's right. I've documented this use case with some sample code